### PR TITLE
Comments: Add actions transitions and notices

### PR DESCRIPTION
**Pull request by Copons**
Introduce in the Comment Management lists both a quick fade out transition for disappearing comments and an undoable global notice in response to actions.

## Changes

- Add `ReactCSSTransitionGroup` to handle a fade out transition on component unmount.
Please note that `ReactCSSTransitionGroup` is deprecated as of React 15.0, and its replacement is tracked in https://github.com/Automattic/wp-calypso/issues/13321.

- Add global notices in response to actions performed on comments.
  - All notices (except "Delete Permanently") feature an Undo button, that simply fires the `setCommentStatus` method with the previous comment status.
  - Undos don't trigger any notice, and are therefore not undoable directly in the notice (they can of course be undone by via comment action).
  - The notices are not stacked together, but performing multiple actions on the same comment (e.g. `approve -> unapprove -> approve -> etc.`) only displays one single notice containing the last performed action.
In other words, we display one notice top for each actioned comment.
  - The notices are set to last 5 seconds before automatically disappearing.
  - The notices are persistent on route changes.
This is to prevent them to disappear when changing Comment Management list.
The side effect, though, is that undoing a notice while outside the Comment Management would trigger the `setCommentStatus` method of the `CommentList` component, which is not mounted anymore.
I'm not exactly sure how to solve this.

### CODE REVIEW

File: client/my-sites/comments/comment-list/index.jsx [hunk 1]
```diff
@@ -31,7 +37,56 @@ export class CommentList extends Component {
...
+			case 'trash':
+				return {
+					type: 'is-error',
+					message: translate( 'Comment moved to trash.' ),
+					options,
+				};
+			case 'delete':
+				return {
+					type: 'is-error',
+					message: translate( 'Comment deleted permanently.' ),
```
Comment (a8ci18n) [inline client/my-sites/comments/comment-list/index.jsx]: Hi! I've found a possible matching string that has already been translated **86** times:
`translate( 'Permanently Delete Comment' )` *ES Score: 8*
See 1 additional suggestion in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/update/comments/actions-transitions-notices)

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

File: client/my-sites/comments/comment-list/index.jsx [hunk 2]
```diff
@@ -31,7 +37,56 @@ export class CommentList extends Component {
...
+			case 'unapproved':
+				return {
+					type: 'is-info',
+					message: translate( 'Comment unapproved.' ),
+					options,
+				};
+			case 'spam':
+				return {
+					type: 'is-warning',
+					message: translate( 'Comment marked as spam.' ),
```
Comment (a8ci18n) [inline client/my-sites/comments/comment-list/index.jsx]: Hi! I've found a possible matching string that has already been translated **45** times:
`translate( 'Comment marked as spam' )` *ES Score: 15*

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

File: client/my-sites/comments/comment-list/index.jsx [hunk 3]
```diff
@@ -31,7 +37,56 @@ export class CommentList extends Component {
...
+			case 'approved':
+				return {
+					type: 'is-success',
+					message: translate( 'Comment approved.' ),
+					options,
+				};
+			case 'unapproved':
+				return {
+					type: 'is-info',
+					message: translate( 'Comment unapproved.' ),
```
Comment (a8ci18n) [inline client/my-sites/comments/comment-list/index.jsx]: Hi! I've found a possible matching string that has already been translated **97** times:
`translate( 'Unapproved' )` *ES Score: 11*
See 2 additional suggestions in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/update/comments/actions-transitions-notices)

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

File: client/my-sites/comments/comment-list/index.jsx [hunk 4]
```diff
@@ -31,7 +37,56 @@ export class CommentList extends Component {
...
+			id: `${ commentId }-newStatus`,
+			isPersistent: true,
+			onClick: () => this.setCommentStatus( commentId, previousStatus ),
+		};
+
+		switch ( newStatus ) {
+			case 'approved':
+				return {
+					type: 'is-success',
+					message: translate( 'Comment approved.' ),
```
Comment (a8ci18n) [inline client/my-sites/comments/comment-list/index.jsx]: Hi! I've found a possible matching string that has already been translated **37** times:
`translate( 'Your comment was approved.' )` *ES Score: 12*

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

File: client/my-sites/comments/comment-list/index.jsx [hunk 5]
```diff
@@ -31,7 +38,57 @@ export class CommentList extends Component {
...
+		const options = {
+			duration: 5000,
+			id: `comment-notice-${ commentId }`,
+			isPersistent: true,
+		};
+
+		if ( 'delete' !== newStatus ) {
+			options.button = translate( 'Undo' );
+			options.onClick = () => this.setCommentStatus( commentId, previousStatus, true );
+		}
```
Comment (dmsnell) [inline client/my-sites/comments/comment-list/index.jsx]: this is exactly how I want undo actions to work!

that being said, this dangling boolean parameter is a bit confusing. is it entirely necessary?

(if we wanted to we could envision a Redux action to dispatch on undo which would contain the payload for setting the comment status. that might be an interesting way to have a global register of undoable actions. not for this PR or this time though)

File: client/my-sites/comments/comment-list/index.jsx [hunk 6]
```diff
@@ -31,7 +38,57 @@ export class CommentList extends Component {
...
+					type: 'is-error',
+					message: translate( 'Comment deleted permanently.' ),
+					options,
+				};
+		}
+	}
+
 	deleteForever = commentId => () => {
+		const notice = this.getNoticeContent( commentId, 'delete', 'trash' );
+		this.props.createNotice( notice.type, notice.message, notice.options );
```
Comment (dmsnell) [inline client/my-sites/comments/comment-list/index.jsx]: note that we can use destructuring here to clean up a bit of noise…

```js
const { type, message, options } = this.getNoticeContent( … );
this.props.createNotice( type, message, options );
```

**Comment by dmsnell** [comment]
@Copons this is really coming along well!

Two points:

 - If the transition group is deprecating, why don't we have an alternative? I'd hate to spend a bunch of time putting it in there if we'll have to soon take it out.

 - I want to caution us about spending too much time writing logic for updating things like comment status because we know that this is _not_ the direction we need to end up with. I'd rather see us use this time to determine the array of Redux actions that we'll need to make the UI work - making them and "faking it until we make it." Some of these already exist, such as the `COMMENTS_CHANGE_STATUS` and we can probably use as-is. Others we may have to make fresh. No worries though, because if no reducer is watching for a new action type then no harm is done. We can scaffold around those "fake it" actions to update local state.

If possible, ideally our test data would still live in `props` but we could use a wrapper to do minimal faking outside. Not sure how much effort it would require now to change but we'll have to change eventually anyway.

```js
const CommentFaker = Wrapped => class extends Component {
	state = {
		comments: {}
	},

	setStatus = ( commentId, status ) => this.setState( { comments: { ...comments, [ commentId ] … );

	render() {
		return (
			<Wrapped { 
				...this.props, 
				comments: this.state.comments,
				setStatus: this.setStatus,
				delete: this.delete,
				trash: this.trash,
				…
			} />
		);
	}
}

export default CommentFaker( CommentManagement );
```

now with this when we get our Redux actions we can do nothing but change out the wrapper. might be handy to keep around too! makes it easy to test or fake data or work with when offline

File: client/blocks/comment-detail/style.scss [hunk 7]
```diff
@@ -18,6 +18,24 @@
 	}
 }
 
+.comment-detail__transition-enter {
+	opacity: 0.01;
```
Comment (gwwar) [inline client/blocks/comment-detail/style.scss]: I'm just curious but why start at `0.01` vs say `0`?

File: client/my-sites/comments/comment-list/index.jsx [hunk 8]
```diff
@@ -3,11 +3,18 @@
  */
 import React, { Component, PropTypes } from 'react';
 import { connect } from 'react-redux';
-import { filter, map } from 'lodash';
+import { localize } from 'i18n-calypso';
+import { filter, keyBy, map } from 'lodash';
+import ReactCSSTransitionGroup from 'react-addons-css-transition-group';
```
Comment (gwwar) [inline client/my-sites/comments/comment-list/index.jsx]: I think we're now OK with the React upgrade, but I recall this component leaving classes on elements in IE11. It'd be worth doing a quick test in that env.

https://github.com/Automattic/wp-calypso/issues/982

File: client/my-sites/comments/comment-list/index.jsx (hunk 5 above)
Comment (gwwar) [inline client/my-sites/comments/comment-list/index.jsx]: @Copons I find that it helps to move bare booleans of long function signatures to an options object, so the reader can have some more context with what's happening. I would also suggest renaming the boolean from `undo` to `showNotice` since we're not attempting to compute any previous state. If we're smart about this I think we can tuck away the notice logic into other middleware.

But as a first step we might do something like this for example:
```
setCommentStatus( commentId, previousStatus, { showNotice: false } )
```

@dmsnell are you suggesting adding an undo queue, where we add an undo action to the queue every time an undoable action is fired?

I've worked with some of these in the past and I find that they undo actions tend to boil down to either computable undos ( I've liked this comment / therefore undo should unlike it ), or undos that need to reset some state, for example restoring post content or unapplying a filter to an image.

I personally would love to see something like this landing, especially for the post editor and image editor.

File: client/blocks/comment-detail/style.scss (hunk 7 above)
Comment (roundhill) [inline client/blocks/comment-detail/style.scss]: I'm curious too :)

Comment (Copons) [inline client/blocks/comment-detail/style.scss]: Fun fact: at some point the transitions weren't working and I couldn't figure out why. The only apparent difference with the docs was that I was using `0` and they `0.01`, so I just tried that.
Eventually I figured out what was my mistake (I was using `ReactCSSTransitionGroup` around the wrong component), but I never went back to that.

Can't seem to find out why they use `0.01`, but for some reason it seems to me that with `0.01` the transitions are a tiny little bit smoother (might very well be my own mind playing tricks on me though).

File: client/my-sites/comments/comment-list/index.jsx (hunk 8 above)
Comment (Copons) [inline client/my-sites/comments/comment-list/index.jsx]: The `ReactCSSTransitionGroup` version we use has explicit timeout props (introduced in https://github.com/facebook/react/pull/4561) that should take care of exactly that issue there.
Considering it's slated to be replaced with `CSSTransitionGroup`, I'd say let's wait for that while we're still behind a feature flag, and then work on a fix if it's really needed.

File: client/my-sites/comments/comment-list/index.jsx [hunk 9]
```diff
@@ -27,53 +34,98 @@ export class CommentList extends Component {
...
+			case 'approved':
+				return createCommentNotice( 'is-success', translate( 'Comment approved.' ), options );
+			case 'unapproved':
+				return createCommentNotice( 'is-info', translate( 'Comment unapproved.' ), options );
+			case 'spam':
+				return createCommentNotice( 'is-warning', translate( 'Comment marked as spam.' ), options );
+			case 'trash':
+				return createCommentNotice( 'is-error', translate( 'Comment moved to trash.' ), options );
+			case 'delete':
+				return createCommentNotice( 'is-error', translate( 'Comment deleted permanently.' ), options );
```
Comment (a8ci18n) [inline client/my-sites/comments/comment-list/index.jsx]: Hi! I've found a possible matching string that has already been translated **86** times:
`translate( 'Permanently Delete Comment' )` *ES Score: 8*
See 1 additional suggestion in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/update/comments/actions-transitions-notices)

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

File: client/my-sites/comments/comment-list/index.jsx [hunk 10]
```diff
@@ -27,53 +34,98 @@ export class CommentList extends Component {
...
+
+		removeCommentNotice( noticeId );
+
+		switch ( newStatus ) {
+			case 'approved':
+				return createCommentNotice( 'is-success', translate( 'Comment approved.' ), options );
+			case 'unapproved':
+				return createCommentNotice( 'is-info', translate( 'Comment unapproved.' ), options );
+			case 'spam':
+				return createCommentNotice( 'is-warning', translate( 'Comment marked as spam.' ), options );
```
Comment (a8ci18n) [inline client/my-sites/comments/comment-list/index.jsx]: Hi! I've found a possible matching string that has already been translated **45** times:
`translate( 'Comment marked as spam' )` *ES Score: 15*

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

File: client/my-sites/comments/comment-list/index.jsx [hunk 11]
```diff
@@ -27,53 +34,98 @@ export class CommentList extends Component {
...
+			options.onClick = () => this.setCommentStatus( commentId, previousStatus, { showNotice: false } );
+		}
+
+		removeCommentNotice( noticeId );
+
+		switch ( newStatus ) {
+			case 'approved':
+				return createCommentNotice( 'is-success', translate( 'Comment approved.' ), options );
+			case 'unapproved':
+				return createCommentNotice( 'is-info', translate( 'Comment unapproved.' ), options );
```
Comment (a8ci18n) [inline client/my-sites/comments/comment-list/index.jsx]: Hi! I've found a possible matching string that has already been translated **97** times:
`translate( 'Unapproved' )` *ES Score: 11*
See 2 additional suggestions in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/update/comments/actions-transitions-notices)

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

File: client/my-sites/comments/comment-list/index.jsx [hunk 12]
```diff
@@ -27,53 +34,98 @@ export class CommentList extends Component {
...
+		if ( 'delete' !== newStatus ) {
+			options.button = translate( 'Undo' );
+			options.onClick = () => this.setCommentStatus( commentId, previousStatus, { showNotice: false } );
+		}
+
+		removeCommentNotice( noticeId );
+
+		switch ( newStatus ) {
+			case 'approved':
+				return createCommentNotice( 'is-success', translate( 'Comment approved.' ), options );
```
Comment (a8ci18n) [inline client/my-sites/comments/comment-list/index.jsx]: Hi! I've found a possible matching string that has already been translated **37** times:
`translate( 'Your comment was approved.' )` *ES Score: 12*

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

File: client/my-sites/comments/comment-list/index.jsx [hunk 13]
```diff
@@ -27,53 +34,98 @@ export class CommentList extends Component {
...
 		if ( ! this.props.comments.length ) {
-			this.setState( { comments: nextProps.comments } );
+			this.setState( { comments: keyBy( nextProps.comments, 'ID' ) } );
+		}
+	}
+
+	showNotice = ( commentId, newStatus, previousStatus ) => {
+		const {
+			createNotice: createCommentNotice,
+			removeNotice: removeCommentNotice,
```
Comment (dmsnell) [inline client/my-sites/comments/comment-list/index.jsx]: personally I'm a fan of changing the `import` so we can leave the props as-is without shadowing concerns. actually I'm not a fan of forbidding shadowing :wink:

if _this_ is the interface (props) then it would make sense to fix the interface and change the periphery

File: client/my-sites/comments/comment-list/index.jsx [hunk 14]
```diff
@@ -27,53 +34,98 @@ export class CommentList extends Component {
...
+			translate,
+		} = this.props;
+
+		const noticeId = `comment-notice-${ commentId }`;
```
Comment (dmsnell) [inline client/my-sites/comments/comment-list/index.jsx]: is this used more than once? any reason to have a separate variable?

File: client/my-sites/comments/comment-list/index.jsx [hunk 15]
```diff
@@ -27,53 +34,98 @@ export class CommentList extends Component {
...
+		const options = {
+			duration: 5000,
+			id: noticeId,
+			isPersistent: true,
+		};
+
+		if ( 'delete' !== newStatus ) {
+			options.button = translate( 'Undo' );
+			options.onClick = () => this.setCommentStatus( commentId, previousStatus, { showNotice: false } );
+		}
```
Comment (dmsnell) [inline client/my-sites/comments/comment-list/index.jsx]: fun fact! we can take advantage of `Object.assign()` to make this declarative…

```js
const options = Object.assign(
	{
		duration: 5000,
		id: `comment-notice-${ commentId }`,
		isPersistent: true,
	},
	'delete' !== newStatus && {
		button: translate( 'Undo' ),
		onClick: () => this.setCommentStatus( … )
	}
);
```

File: client/my-sites/comments/comment-list/index.jsx (hunk 9 above)
Comment (dmsnell) [inline client/my-sites/comments/comment-list/index.jsx]: fun fact! if you find this repetitive, we could also modularize…

```js
const [ type, message ] = get( {
	approved: [ 'is-success', translate( … ) ],
	unapproved: …
	…
}, newStatus, [ null, null ] );

if ( ! type ) {
	return;
}

createCommentNotice( type, message, options );
```

File: client/my-sites/comments/comment-list/index.jsx [hunk 16]
```diff
@@ -27,53 +34,96 @@ export class CommentList extends Component {
...
 
+	showNotice = ( commentId, newStatus, previousStatus ) => {
+		const { translate } = this.props;
+
+		const [ type, message ] = get( {
+			approved: [ 'is-success', translate( 'Comment approved.' ) ],
+			unapproved: [ 'is-info', translate( 'Comment unapproved.' ) ],
+			spam: [ 'is-warning', translate( 'Comment marked as spam.' ) ],
+			trash: [ 'is-error', translate( 'Comment moved to trash.' ) ],
+			'delete': [ 'is-error', translate( 'Comment deleted permanently.' ) ],
```
Comment (a8ci18n) [inline client/my-sites/comments/comment-list/index.jsx]: Hi! I've found a possible matching string that has already been translated **86** times:
`translate( 'Permanently Delete Comment' )` *ES Score: 8*
See 1 additional suggestion in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/update/comments/actions-transitions-notices)

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

File: client/my-sites/comments/comment-list/index.jsx [hunk 17]
```diff
@@ -27,53 +34,96 @@ export class CommentList extends Component {
...
 		}
 	}
 
+	showNotice = ( commentId, newStatus, previousStatus ) => {
+		const { translate } = this.props;
+
+		const [ type, message ] = get( {
+			approved: [ 'is-success', translate( 'Comment approved.' ) ],
+			unapproved: [ 'is-info', translate( 'Comment unapproved.' ) ],
+			spam: [ 'is-warning', translate( 'Comment marked as spam.' ) ],
```
Comment (a8ci18n) [inline client/my-sites/comments/comment-list/index.jsx]: Hi! I've found a possible matching string that has already been translated **45** times:
`translate( 'Comment marked as spam' )` *ES Score: 15*

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

File: client/my-sites/comments/comment-list/index.jsx [hunk 18]
```diff
@@ -27,53 +34,96 @@ export class CommentList extends Component {
...
+			this.setState( { comments: keyBy( nextProps.comments, 'ID' ) } );
 		}
 	}
 
+	showNotice = ( commentId, newStatus, previousStatus ) => {
+		const { translate } = this.props;
+
+		const [ type, message ] = get( {
+			approved: [ 'is-success', translate( 'Comment approved.' ) ],
+			unapproved: [ 'is-info', translate( 'Comment unapproved.' ) ],
```
Comment (a8ci18n) [inline client/my-sites/comments/comment-list/index.jsx]: Hi! I've found a possible matching string that has already been translated **97** times:
`translate( 'Unapproved' )` *ES Score: 11*
See 2 additional suggestions in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/update/comments/actions-transitions-notices)

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

File: client/my-sites/comments/comment-list/index.jsx [hunk 19]
```diff
@@ -27,53 +34,96 @@ export class CommentList extends Component {
...
-			this.setState( { comments: nextProps.comments } );
+			this.setState( { comments: keyBy( nextProps.comments, 'ID' ) } );
 		}
 	}
 
+	showNotice = ( commentId, newStatus, previousStatus ) => {
+		const { translate } = this.props;
+
+		const [ type, message ] = get( {
+			approved: [ 'is-success', translate( 'Comment approved.' ) ],
```
Comment (a8ci18n) [inline client/my-sites/comments/comment-list/index.jsx]: Hi! I've found a possible matching string that has already been translated **37** times:
`translate( 'Your comment was approved.' )` *ES Score: 12*

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).
